packages/@lwc/integration-types/src/decorators/wire.ts

```diff
@@ -0,0 +1,142 @@
+/*
+ * Copyright (c) 2024, Salesforce, Inc.
+ * All rights reserved.
+ * SPDX-License-Identifier: MIT
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
+ */
+
+import { WireAdapter } from '@lwc/engine-core';
+import { LightningElement, WireAdapterConstructor, wire } from 'lwc';
+
+type WireConfig = { config: 'config' };
+type WireValue = { value: 'value' };
+type WireContext = { context: 'context' };
+type DeepConfig = { deep: { config: number } };
+
+const config: WireConfig = { config: 'config' };
+
+// `class C implements A` validates that the class instance matches type A
+// `const C = class C {} satisfies B` validates that the class constructor matches type B
+const FakeWireAdapter = class FakeWireAdapter implements WireAdapter<WireConfig, WireContext> {
+    constructor(private cb: (value: WireValue) => void) {}
+    update(_cfg: WireConfig, _ctx: WireContext) {}
+    connect() {}
+    disconnect() {}
+} satisfies WireAdapterConstructor<WireConfig, WireValue, WireContext>;
+
+const DeepConfigAdapter = class DeepConfigAdapter implements WireAdapter<DeepConfig, WireValue> {
+    constructor(private cb: (value: WireValue) => void) {}
+    update(_cfg: DeepConfig) {}
+    connect() {}
+    disconnect() {}
+} satisfies WireAdapterConstructor<DeepConfig, WireValue>;
+
+// @ts-expect-error bare decorator cannot be used
+wire(FakeWireAdapter, { config: 'config' })();
+
+// @ts-expect-error decorator cannot be used on classes
+@wire(FakeWireAdapter, { config: 'config' })
+export class InvalidDecoratorContexts extends LightningElement {
+    // @ts-expect-error decorator cannot be used on getters
```

Comment:
As mentioned in https://github.com/salesforce/lwc/issues/4428#issuecomment-2266140289 we should allow this.